chips/sam4l/src/usart.rs

```diff
@@ -549,6 +671,187 @@ impl hil::uart::UARTAdvanced for USART {
+        regs.rhr.get() as u8
+    }
+
+    /// Don't call this to use the HW chip select pin on the USART (RTS).
+    fn specify_chip_select(&self, cs: Self::ChipSelect) {
+        self.spi_chip_select.set(Some(cs));
+    }
+
+    /// Returns the actual rate set
+    fn set_rate(&self, rate: u32) -> u32 {
+        self.set_baud_rate(rate);
+
+        // Calculate what rate will actually be
+        let system_frequency = unsafe { pm::get_system_frequency() };
+        let cd = system_frequency / rate;
+        system_frequency / cd
+    }
+
+    fn get_rate(&self) -> u32 {
+        let regs: &mut USARTRegisters = unsafe { mem::transmute(self.registers) };
+        let system_frequency = unsafe { pm::get_system_frequency() };
+        let cd = regs.brgr.get() & 0xFFFF;
+        system_frequency / cd
+    }
+
+    fn set_clock(&self, polarity: hil::spi::ClockPolarity) {
+        let regs: &mut USARTRegisters = unsafe { mem::transmute(self.registers) };
+        let mode = regs.mr.get();
+
+        match polarity {
+            hil::spi::ClockPolarity::IdleLow => {
+                regs.mr.set(mode & !(1 << 16));
+            }
+            hil::spi::ClockPolarity::IdleHigh => {
+                regs.mr.set(mode | (1 << 16));
+            }
+        }
+    }
+
+    fn get_clock(&self) -> hil::spi::ClockPolarity {
+        let regs: &mut USARTRegisters = unsafe { mem::transmute(self.registers) };
+        let mode = regs.mr.get();
+
+        match mode & (1 << 16) {
+            0 => hil::spi::ClockPolarity::IdleLow,
+            _ => hil::spi::ClockPolarity::IdleHigh,
+        }
+    }
+
+    fn set_phase(&self, phase: hil::spi::ClockPhase) {
+        let regs: &mut USARTRegisters = unsafe { mem::transmute(self.registers) };
+        let mode = regs.mr.get();
+
+        match phase {
+            hil::spi::ClockPhase::SampleLeading => {
+                regs.mr.set(mode | (1 << 8));
+            }
+            hil::spi::ClockPhase::SampleTrailing => {
+                regs.mr.set(mode & !(1 << 8));
+            }
+        }
+    }
+
+    fn get_phase(&self) -> hil::spi::ClockPhase {
+        let regs: &mut USARTRegisters = unsafe { mem::transmute(self.registers) };
+        let mode = regs.mr.get();
+
+        match mode & (1 << 8) {
+            0 => hil::spi::ClockPhase::SampleLeading,
+            _ => hil::spi::ClockPhase::SampleTrailing,
+        }
+    }
+
+    // These two functions determine what happens to the chip
+    // select line between transfers. If hold_low() is called,
+    // then the chip select line is held low after transfers
+    // complete. If release_low() is called, then the chip select
+    // line is brought high after a transfer completes. A "transfer"
+    // is any of the read/read_write calls. These functions
+    // allow an application to manually control when the
+    // CS line is high or low, such that it can issue multi-byte
+    // requests with single byte operations.
+    fn hold_low(&self) {}
```

Comment:
Well, because I was hoping they would go away. They are only needed in byte-by-byte operation, which we probably don't want to encourage. If they (or things like them) are needed, they really should be options to `read_write_bytes()`.